app/scripts/modules/core/src/pipeline/config/triggers/TriggersPageContent.tsx

```diff
@@ -0,0 +1,144 @@
+import * as React from 'react';
+
+import { extend, findIndex } from 'lodash';
+
+import { Application } from 'core/application';
+import { ArtifactReferenceService } from 'core/artifact/ArtifactReferenceService';
+import { IExpectedArtifact, IPipeline, ITrigger } from 'core/domain';
+import { HelpField } from 'core/help';
+import { PipelineConfigValidator } from 'core/pipeline';
+import { CheckboxInput, FormField } from 'core/presentation';
+import { Registry } from 'core/registry';
+import { SETTINGS } from 'core/config/settings';
+import { PipelineRoles } from './PipelineRoles';
+import { Trigger } from './Trigger';
+
+export interface ITriggersPageContentProps {
+  application: Application;
+  pipeline: IPipeline;
+  updatePipelineConfig: (changes: Partial<IPipeline>) => void;
+}
+
+export function TriggersPageContent(props: ITriggersPageContentProps) {
+  const showProperties = SETTINGS.feature.quietPeriod || SETTINGS.feature.managedServiceAccounts;
+  const { pipeline, application, updatePipelineConfig } = props;
+
+  function updateRoles(roles: any[]): void {
+    updatePipelineConfig({ roles });
+  }
+
+  function addTrigger(): void {
+    const triggerTypes = Registry.pipeline.getTriggerTypes();
+    let newTrigger: ITrigger = { enabled: true, type: null };
+    if (triggerTypes.length === 1) {
+      newTrigger = { enabled: true, type: triggerTypes[0].key };
+    }
+    updatePipelineConfig({ triggers: [...pipeline.triggers, newTrigger] });
+  }
+
+  function removeTrigger(triggerIndex: number): void {
+    const triggers = pipeline.triggers.slice(0);
+    triggers.splice(triggerIndex, 1);
+    updatePipelineConfig({ triggers });
+  }
+
+  function updateTrigger(index: number, changes: Partial<ITrigger>) {
+    const triggers = pipeline.triggers.slice(0);
+    extend(triggers[index], changes);
+    PipelineConfigValidator.validatePipeline(pipeline);
+    updatePipelineConfig({ triggers });
+    if (SETTINGS.feature['artifactsRewrite']) {
+      removeUnusedExpectedArtifacts(pipeline);
+    }
+  }
+
+  // Expected Artifacts
+  function updateExpectedArtifacts(expectedArtifacts: IExpectedArtifact[]) {
+    updatePipelineConfig({ expectedArtifacts });
+  }
+
+  function removeUnusedExpectedArtifacts(pipelineParam: IPipeline) {
+    // remove unused expected artifacts from the pipeline
+    const expectedArtifacts: IExpectedArtifact[] = [...pipelineParam.expectedArtifacts];
```

Comment:
any reason we need to spread this into a new array? looks like we're just iterating through it with `forEach` but I might be missing something